resources/type-system/flow-analysis.md

```diff
@@ -0,0 +1,713 @@
+  - `VM = VariableModel(declared, promoted, tested, assigned, unassigned, captured)`
+  - `promoted = S::l` or (`promoted = []` and `declared = S`)
+
+Policy:
+  - We say that at type `T` is a type of interest for a variable `x` in a set of
+    tested types `tested` if `tested` contains a type `S` such that `T` is `S`,
+    or `T` is **NonNull(`S`)**.
+
+  - We say that a variable `x` is promotable via type test with type `T` given
+    variable model `VM` if
+    - `VM = VariableModel(declared, promoted, tested, assigned, unassigned, captured)`
+    - and `captured` is false
+    - and `S` is the current type of `x` in `VM`
+    - and not `S <: T`
+    - and `T <: S` or (`S` is `X extends R` and `T <: R`) or (`S` is `X & R` and
+      `T <: R`)
+
+  - We say that a variable `x` is promotable via initialization given variable
+    model `VM` if:
+    - `VM = VariableModel(declared, promoted, tested, assigned, unassigned, captured)`
+    - and `captured` is false
+    - and `promoted` is empty
+    - and `x` is declared with no explicit type and no initializer
+    - and `assigned` is false and `unassigned` is true
+
+  - We say that a variable `x` is promotable via assignment of an expression of
+    type `T` given variable model `VM` if
+    - `VM = VariableModel(declared, promoted, tested, assigned, unassigned, captured)`
+    - and `captured` is false
+    - and `S` is the current type of `x` in `VM`
+    - and `T <: S` and not `S <: T`
+    - and `T` is a type of interest for `x` in `tested`
+
+  - We say that a variable `x` is demotable via assignment of an expression of
+    type `T` given variable model `VM` if
+    - `VM = VariableModel(declared, promoted, tested, assigned, unassigned, captured)`
+    - and `captured` is false
+    - and declared::promoted contains a type `S` such that `T` is `S` or `T` is
+      **NonNull(`S`)**.
+
+Definitions:
+
+- `assign(x, E, M)` where `x` is a local variable, `E` is an expression of
+  inferred type `T`, and `M = FlowModel(r, VI)` is the flow model for `E` is
+  defined to be `FlowModel(r, VI[x -> VM])` where:
+    - `VI(x) = VariableModel(declared, promoted, tested, assigned, unassigned, captured)`
+    - if `captured` is true then:
+      - `VM = VariableModel(declared, promoted, tested, true, false, captured)`.
+    - otherwise if `x` is promotable via initialization given `VM` then
+      - `VM = VariableModel(declared, [T], tested, true, false, captured)`.
+    - otherwise if `x` is promotable via assignment of `E` given `VM`
+      - `VM = VariableModel(declared, T::promoted, tested, true, false, captured)`.
+    - otherwise if `x` is demotable via assignment of `E` given `VM`
+      - `VM = VariableModel(declared, demoted, tested, true, false, captured)`.
+      - where `previous` is the suffix of `promoted` starting with the first type
+        `S` such that `T <: S`, and:
+        - if `S`is nullable and if `T <: Q` where `Q` is **NonNull(`S`)** then
```

Comment:
I believe the "NonNull" part of this is not yet implemented.  Tracking in https://github.com/dart-lang/sdk/issues/40011.